document_similarity/microservice/library/postgresql.py

```diff
@@ -138,28 +139,13 @@ def retrieve_similarities(self, doc_id, k=5):
         ORDER BY similarity_score DESC;
         """
         similarity_list = self.execute(statement, (doc_id,))
-        result_indices = [entry['document2_id'] for entry in similarity_list[:k]]
-        result = [(entry['document2_id'], entry['similarity_score']) for entry in similarity_list[:k]]
+        if offset > len(similarity_list):
+            raise Exception("The number of documents in the database is lower than the 'offset' parameter.")
+        limit = min(offset + k, len(similarity_list))
+        result_indices = [entry['document2_id'] for entry in similarity_list[offset:limit]]
+        result = [(entry['document2_id'], entry['similarity_score']) for entry in similarity_list[offset:limit]]
```

Comment:
I agree that this is very useful however this should not be implemented this way, since you are extracting all of the similarities from the database into memory and then doing calculations on that. Since the number of results can be huge, this could cause bad performances on the microservice.
Currently we do not have that many similarities in the table and therefore it would be hard to notice, but if we had 100000 of similiarities or more, there would be noticeable differences in speed and performance.

You can achieve the same results with postgres avoiding extracting all similarities from the database. You can find some help here: 
https://www.postgresql.org/docs/8.0/queries-limit.html

Reply:
I agree with Samo - loading all of the data will cause performance problems. In addition, the current implementation of the `execute` function is intended for retrieving smaller amounts of data. For bigger amounts, we would need to use a different approach.

What Samo is a good solution.